include/rmm/device_buffer.hpp

```diff
@@ -104,6 +105,7 @@ class device_buffer {
                          mr::device_memory_resource* mr = mr::get_current_device_resource())
     : _stream{stream}, _mr{mr}
   {
+    cuda_set_device_raii dev{_device};
     allocate_async(size);
```

Comment:
**question:** Should the setting of the current device live inside `allocate/deallocate` rather than it being the responsibility of the caller to ensure the device is correct? Or, is this deliberate because we might want more than just the allocate call to occur with the same device active and this approach avoids excessive device switching?

Reply:
Yeah, it's deliberate. I wanted to put it in `allocate_async`/`deallocate_async`, but those calls are often made in places where the correct device is also needed for other operations, and we don't want to `cuda_set_device_raii` multiple times.  There are also places such as `resize / shrink_to_fit` where a new `device_buffer` is created and we want that to happen with the original device active, but inside it we call `allocate_async` and that would cause redundant current device checking.

Reply:
I think what I've arrived at is that in order to minimize device switching, we want to do it at the highest level in `device_buffer` possible, which means the public API functions (when necessary). For the same reason, we assume the user has set the device before constructing the `device_buffer`, and we just store the ID at that stage.